bevy_lint/src/bin/driver.rs

```diff
@@ -2,16 +2,51 @@
 #![feature(rustc_private)]
 
 extern crate rustc_driver;
+extern crate rustc_session;
 extern crate rustc_span;
 
+use std::process::ExitCode;
+
 use bevy_lint::BevyLintCallback;
-use rustc_span::ErrorGuaranteed;
+use rustc_driver::{catch_with_exit_code, init_rustc_env_logger, install_ice_hook, RunCompiler};
+use rustc_session::{config::ErrorOutputType, EarlyDiagCtxt};
+
+const BUG_REPORT_URL: &str = "https://github.com/TheBevyFlock/bevy_cli/issues";
+
+fn main() -> ExitCode {
+    // Setup a diagnostic context that can be used for error messages.
+    let early_dcx = EarlyDiagCtxt::new(ErrorOutputType::default());
+
+    // Setup `rustc`'s builtin `tracing` logger.
+    init_rustc_env_logger(&early_dcx);
+
+    // "ICE" stands for Internal Compiler Error. An ICE hook is a special type of panic handler
+    // that dumps an absurd amount of data when the driver panics. We take advantage of this, but
+    // override the default bug report URL so that users bother us, not Rust compiler devs. :)
+    install_ice_hook(BUG_REPORT_URL, |dcx| {
+        dcx.handle()
+            .note("This is likely a bug with `bevy_lint`, not `rustc` or `cargo`.");
+    });
+
+    // Run the passed closure, but catch any panics and return the respective exit code.
+    let exit_code = catch_with_exit_code(move || {
+        // Get the arguments passed through the CLI. This is equivalent to `std::env::args()`, but
+        // it returns a `Result` instead of panicking.
+        let mut args = rustc_driver::args::raw_args(&early_dcx)?;
+
+        // The arguments are formatted as `[DRIVER_PATH, RUSTC_PATH, ARGS...]`. We skip the driver
+        // path so that `RunCompiler` just sees `rustc`'s path.
+        args.remove(0);
+
+        println!("{:?}", args);
```

Comment:
Note to self: this was added while debugging and should definitely be removed before this is merged.

Reply:
> I'm currently debugging an issue with this where serde_derive fails to compile (I believe it's the rustc_driver::args::raw_args() call), so I'll update this when I fix it. :)

Turns out this issue was due to this `println!()` statement. The error I was getting was:

```bash
error: extern location for proc_macro2 does not exist: ~/dev/bevy/bevy_cli/target/debug/deps/proc_macro2-3e9157d6fe8cf67e
  --> ~/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde_derive-1.0.210/src/lib.rs:65:1
   |
65 | extern crate proc_macro2;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^

error: extern location for quote does not exist: ~/dev/bevy/bevy_cli/target/debug/deps/quote-7e885e866937aa79
  --> ~/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde_derive-1.0.210/src/lib.rs:66:1
   |
66 | extern crate quote;
   | ^^^^^^^^^^^^^^^^^^^

error: extern location for syn does not exist: ~/dev/bevy/bevy_cli/target/debug/deps/syn-7bbfee08b5f907af
  --> ~/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde_derive-1.0.210/src/lib.rs:67:1
   |
67 | extern crate syn;
   | ^^^^^^^^^^^^^^^^^

error: could not compile `serde_derive` (lib) due to 3 previous errors
```

The reason I believe this was happening was because `proc_macro2` has a `build.rs` that [runs](https://github.com/dtolnay/proc-macro2/blob/9c1d3eb1b6bddc5b86522bf3af98c2fb3de8e30a/build.rs#L135) `rustc` [twice](https://github.com/dtolnay/proc-macro2/blob/9c1d3eb1b6bddc5b86522bf3af98c2fb3de8e30a/build.rs#L194), and parses the output.

This `println!()` was getting in the way of the parsed output, so it failed and caused a series build error. Removing this line causes the build to succeed.

The more you know 🌟